src/array_like.rs

```diff
@@ -0,0 +1,111 @@
+//! TODO
+
+use crate::{
+    ndarray::{Array, Array1, ArrayView, Axis, Dimension},
+    Element, PyReadonlyArray,
+};
+use ndarray::{Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{exceptions::PyValueError, FromPyObject, PyAny, PyResult};
+
+/// TODO
+#[derive(Debug)]
+pub struct PyArrayLike<'py, T, D>(ArrayLike<'py, T, D>)
+where
+    T: Element,
+    D: Dimension;
+
+#[derive(Debug)]
+enum ArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    Reference(PyReadonlyArray<'py, T, D>),
+    Converted(Array<T, D>),
+}
+
+impl<'py, T, D> PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    /// TODO
+    pub fn into_owned_array(self) -> Array<T, D> {
+        match self.0 {
+            ArrayLike::Converted(x) => x,
+            ArrayLike::Reference(x) => x.to_owned_array(),
+        }
+    }
+
+    /// TODO
+    pub fn as_view(&self) -> ArrayView<T, D> {
+        match &self.0 {
+            ArrayLike::Converted(x) => x.view(),
+            ArrayLike::Reference(x) => x.as_array(),
+        }
+    }
+
+    /// TODO
+    pub fn is_reference(&self) -> bool {
+        match &self.0 {
+            ArrayLike::Reference(_) => true,
+            ArrayLike::Converted(_) => false,
+        }
+    }
+}
+
+impl<'py, T, D> FromPyObject<'py> for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+    Vec<T>: FromPyObject<'py>,
+{
+    fn extract(ob: &'py PyAny) -> PyResult<Self> {
+        if let Ok(array) = ob.extract::<PyReadonlyArray<T, D>>() {
+            return Ok(PyArrayLike(ArrayLike::Reference(array)));
+        }
+
+        if D::NDIM.unwrap_or(1) == 1 {
+            if let Ok(array) = ob.extract::<Vec<T>>() {
+                let res = Array1::from(array)
+                    .into_dimensionality()
+                    .map_err(|_| PyValueError::new_err("Dimension incompatible"))?;
```

Comment:
Can this really fail? If `D::NDIM == Some(1)`, it should not be possible because the dimensionality matches. If `D::NDIM == None`, it should not be possible because the dimensionality does not matter.

Reply:
Usually, it should not fail. But I was not sure about custom implementations of `Dimension`, so I thought "better safe than sorry".

Reply:
`Dimension` is a sealed trait so there can be no custom implementations outside of `ndarray`. Hence, please just unwrap here.